Refactor search functionality

OliverRhyme opened the pull request:
@joshuadeguzman 
Refactors the search functionality with the flexibility to easily integrate Jobs search.
Edit: ~~Add Developer loading~~

OliverRhyme commented on lib/features/devboard/devboard_page.dart:
```diff
@@ -14,28 +12,69 @@ class DevboardPage extends StatefulWidget {
+}
+
+// Can also be used on jobs page
+class DevProgress extends StatelessWidget {
```
Improve the design of this

joshuadeguzman replied:
Please create an issue ticket for this. Thank you @OliverRhyme

joshuadeguzman commented:
Hello @OliverRhyme,

Few things from me,
- Please add the screenshots for the changes made (preferably GIF or video?)
- Please rebase to the latest commit

Thanks for this.

joshuadeguzman commented on .all-contributorsrc:
```diff
@@ -426,6 +426,24 @@
       "contributions": [
         "content"
       ]
+    },
```
Please rebase to the latest master branch :D

OliverRhyme replied:
Done rebasing thank you.

OliverRhyme commented:
> Hello @OliverRhyme,
> 
> Few things from me,
> 
> * Please add the screenshots for the changes made (preferably GIF or video?)
> * Create issue tickets for the changes you made, eg. create an issue for refactoring search
> * Please rebase to the latest commit
> 
> Thanks for this.

I am able to run the page on my github fork at [https://oliverrhyme.github.io/devs](url)

joshuadeguzman commented on lib/core/widgets/components/search_bar.dart:
```diff
@@ -44,7 +44,7 @@ class _SearchBarState extends State<SearchBar> {
           decoration: InputDecoration(
             border: InputBorder.none,
-            hintText: 'Search name, Github usernames...',
+            hintText: "Who are you looking for?",
```
I recommend not changing this one for now.

joshuadeguzman commented on lib/features/dashboard/dashboard_model.dart:
```diff
@@ -1,55 +1,61 @@
+          ? devsList
+          : devsList
+              .where((dev) =>
+                  // TODO: Improve null safety efficiency, maybe utilizing extension function
```
Kindly create an issue ticket to migrate this to a utility or extension.

joshuadeguzman commented on lib/features/dashboard/dashboard_model.dart:
```diff
@@ -1,55 +1,61 @@
 
   void setSelectedPageIndex(int index) {
     selectedPageIndex = index;
+    // TODO: Add a switch statement to handle stream switching when selecting page
```
Are you adding it on a follow up PR?

OliverRhyme replied:
Just an idea for whoever works on jobs page in the future.

joshuadeguzman commented on lib/features/dashboard/dashboard_page.dart:
```diff
@@ -28,28 +27,28 @@ class DashboardPageState extends State<DashboardPage> {
-            create: (_) => DevBoardModel(devsRepository),
-          ),
-        ],
+      // MultiProvider(
```
Let's stick with MultiProvider to avoid unnecessary changes when adding or removing more providers.

OliverRhyme replied:
Noted thank you.

joshuadeguzman commented on lib/features/devboard/devboard_page.dart:
```diff
@@ -14,28 +12,69 @@ class DevboardPage extends StatefulWidget {
+  @override
+  void initState() {
+    super.initState();
+    devBoardModel =  Provider.of(context, listen: false);
```
`context` here is not available upon initialising the widget before first render.

Please see [this](https://stackoverflow.com/questions/60343605/how-to-use-data-from-provider-during-initstate-in-flutter-application) or maintain the initialisation in `didChangeDependencies`

OliverRhyme replied:
Its weird that following the StackOverflow instead breaks the code, but the this method works and also the `didChangeDependencies` method. Anyway will revert back to `didChangeDependencies`, thank you.

joshuadeguzman commented on lib/main.dart:
```diff
@@ -37,12 +40,13 @@ class _DevsAppState extends State<DevsApp> {
-        Provider(
-          create: (_) => devsRepository,
-        ),
+        // No need to expose the sources
```
Please ignore this for now.

joshuadeguzman reviewed (changes requested):
Hello @OliverRhyme,

Thanks for the PR. 🎉 

Kindly review the comments. For the meantime, let's ignore issues that are not related to the PR, eg. providing repository and data source, to keep the PR small as possible.

cc: @ram231  @Jansalvador1445

OliverRhyme commented:
> Hello @OliverRhyme,
> 
> Thanks for the PR. 🎉
> 
> Kindly review the comments. For the meantime, let's ignore issues that are not related to the PR, eg. providing repository and data source, to keep the PR small as possible.
> 
> cc: @ram231 @Jansalvador1445

I've just refactored, as it is needed for the search to work properly

OliverRhyme commented:
Separated dev loading indicator into another pull request

OliverRhyme commented:
Hi po @joshuadeguzman @Jansalvador1445 pa review po. I've already addressed some of the concerns.

Note: This will overwrite the #282 since its not compatible, but I have my own implementation with the follow up PR which can easily be improved.

ram231 commented on lib/utils/utils.dart:
```diff
@@ -0,0 +1,5 @@
+extension CaseInsensitive on String {
+  bool containsIgnoreCase(String other) {
+    return this == null ? false : this.toLowerCase().contains(other.toLowerCase());
```
It can be rewritten as `this?.toLowerCase()?.contains(other.toLowerCase()) ?? false;` but it does the same thing. You can practice it on your own to make less ternary statement less complicated. well anyways, LGTM.

OliverRhyme replied:
Yes I forgot to improve this when im debugging the null safety, anyway I have a follow-up PR. Thank you.

ram231 reviewed (approved):
![LGTM](https://blog.alexv525.com/wp-content/uploads/2020/08/LGTM.png)

joshuadeguzman commented:
@all-contributors please add @OliverRhyme  for code, bug
